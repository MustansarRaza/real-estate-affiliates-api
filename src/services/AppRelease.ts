import { AppRelease } from "affiliates-api/models";
import { Op } from "sequelize";

export interface AppVariablesDTO {
  min_version: string;
}
export interface SettingsDTO {
  version: string;
  appVariables: AppVariablesDTO;
}
class AppReleaseService {
  /**
   * Returns mobile application settings
   */
  async get(appName: string): Promise<SettingsDTO | null> {
    const release = await AppRelease.findOne({
      attributes: ["version", "appVariables"],
      where: {
        appName: {
          [Op.eq]: appName,
        },
      },
    });

    return release ? this.serialize(release) : null;
  }
  /**
   * Serializes a queried {@see AppRelease} model instance into a DTO.
   */
  private serialize(release: AppRelease): SettingsDTO {
    return {
      version: release.version.toString(),
      appVariables: {
        min_version: release.appVariables!.min_version,
      },
    };
  }
}
export default AppReleaseService;
