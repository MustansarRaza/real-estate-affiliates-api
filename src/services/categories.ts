import { ListingStatusKey, PlatformKey } from "@pf/taxonomies";
import {
  Address,
  AgencyConfiguration,
  Category,
  Listing,
  ListingStatus,
  Location,
  Platform,
  popularCategoryKeyToIndex,
  Project,
  ProjectAffiliateVisibilityID,
  ProjectStatus,
} from "affiliates-api/models";
import { col, fn, Op } from "sequelize";

interface CategoryAvailabilityDTO {
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

interface CategoryCityAvailabilityDTO {
  cityID: number;
  unitsCount: number;
}

interface CategoryDTO {
  id: number;
  name: string;
  nameSingular: string;
  orderIndex: number;
  slug: string;
  parentID: number | null;
  popularityIndex: number;
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

/**
 * Provides access to categories.
 */
class CategoriesService {
  /**
   * Gets a list of all available categories.
   */
  async list(agencyID: number | null): Promise<CategoryDTO[]> {
    let platformId = null;

    if (agencyID) {
      const agencyConfig = await AgencyConfiguration.findOne({
        attributes: ["platformId"],
        where: { agencyId: { [Op.eq]: agencyID } },
        raw: true,
      });
      platformId = agencyConfig?.platformId;
    }
    if (!platformId) {
      const platform = await Platform.findOne({
        attributes: ["id"],
        where: { platformKey: { [Op.eq]: PlatformKey.ZAMEEN } },
        raw: true,
      });
      platformId = platform.id;
    }

    const categories = await Category.findAll({
      where: { platformId: { [Op.eq]: platformId } },
      order: [
        ["type_order", "ASC"],
        ["type_id", "ASC"],
      ],
    });

    const categoryCitiesAndProjectsMap = await this.getCategoryCitiesAndProjectsMap(agencyID);

    return categories.map((category: Category) => ({
      id: category.id,
      name: category.name,
      nameSingular: category.nameSingular,
      orderIndex: category.orderIndex,
      slug: category.slug,
      parentID: category.parentID,
      popularityIndex: popularCategoryKeyToIndex(category.key),
      availableInCities: categoryCitiesAndProjectsMap[category.id]?.availableInCities || [],
      availableInProjects: categoryCitiesAndProjectsMap[category.id]?.availableInProjects || [],
    }));
  }

  /**
   * Gets a map of cities and Projects against a category in which
   * that category is available.
   */
  private async getCategoryCitiesAndProjectsMap(
    agencyID: number | null
  ): Promise<{
    [propertTypeId: number]: CategoryAvailabilityDTO;
  }> {
    const categoriesAvailablityResult = await Listing.findAll({
      attributes: [[fn("COUNT", "id"), "Count"]],
      include: [
        {
          model: ListingStatus,
          attributes: [],
          where: {
            key: {
              [Op.eq]: ListingStatusKey.AVAILABLE,
            },
          },
          required: true,
        },
        {
          model: Category,
          attributes: ["id"],
          required: true,
        },
        {
          model: Project,
          required: true,
          attributes: ["id"],
          where: {
            status: {
              [Op.eq]: ProjectStatus.ON,
            },
            affiliateVisibility: {
              [Op.eq]: ProjectAffiliateVisibilityID.VISIBLE,
            },
            ...(agencyID && {
              agencyID: {
                [Op.eq]: agencyID,
              },
            }),
          },
          include: [
            {
              model: Address,
              required: true,
              attributes: [],
              include: [
                {
                  model: Location,
                  attributes: ["id"],
                  required: true,
                },
              ],
            },
          ],
        },
      ],
      group: [
        col("Category.type_id"),
        col("Project.Address.Location.title"),
        col("Project.project_id"),
      ],
      raw: true,
    });

    const categoryCitiesAndProjectsMap: { [propertTypeId: number]: CategoryAvailabilityDTO } = {};
    categoriesAvailablityResult.forEach((row: any) => {
      const propertyTypeId = row["Category.id"];
      const projectID = row["Project.id"];
      const cityID = row["Project.Address.Location.id"];
      const unitsCount = row["Count"];

      if (categoryCitiesAndProjectsMap.hasOwnProperty(propertyTypeId)) {
        categoryCitiesAndProjectsMap[propertyTypeId].availableInProjects.push(projectID);
        const index = categoryCitiesAndProjectsMap[propertyTypeId].availableInCities
          .map((city: CategoryCityAvailabilityDTO) => city.cityID)
          .indexOf(cityID);
        if (index === -1) {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities.push({
            cityID,
            unitsCount,
          });
        } else {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities[
            index
          ].unitsCount += unitsCount;
        }
      } else {
        categoryCitiesAndProjectsMap[propertyTypeId] = {
          availableInProjects: [projectID],
          availableInCities: [{ cityID, unitsCount }],
        };
      }
    });

    return categoryCitiesAndProjectsMap;
  }
}

export default CategoriesService;
