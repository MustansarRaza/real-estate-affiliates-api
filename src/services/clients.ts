import { SourcePlatform } from "@pf/taxonomies";
import { AppError } from "@pf/utils";
import {
  AffiliateContact,
  Client,
  ClientPhoneNumber,
  ClientReferralType,
  ClientStatusID,
} from "affiliates-api/models";
import { ISO8601DateTime } from "affiliates-api/types";
import { parsePhoneNumberFromString, PhoneNumber } from "libphonenumber-js";
import { Op, Sequelize, WhereOptions } from "sequelize";
import LocationsService from "./locations";
import MeService from "./me";

export interface ClientPhoneNumberDTO {
  id: number;
  value: string;
}

export interface ClientDTO {
  id: number;
  name: string;
  email: string | null;
  phoneNumbers: ClientPhoneNumberDTO[];
  address: string | null;
  locationID: number | null;
  comment: string | null;
  createdAt: ISO8601DateTime;
  updatedAt: ISO8601DateTime;
}

export interface CreateClientPhoneNumberDTO {
  value: string;
}

export interface CreateClientDTO {
  name: string;
  email: string;
  phoneNumbers: CreateClientPhoneNumberDTO[];
  address: string | null;
  locationID: number | null;
  comment: string | null;
  source?: SourcePlatform;
}

interface OtherServices {
  me: MeService;
  locations: LocationsService;
}

interface BySSOIDLookup {
  ssoID: string;
}

type ByNamePhoneNumberOrEmailLookup = RequireAtLeastOne<{
  name?: string | null;
  phoneNumber?: string | null;
  email?: string | null;
}>;

/**
 * Provides access to clients.
 *
 * Clients are essentially contacts. They represent a person which can
 * be assigned to one or more leads.
 */
class ClientsService {
  constructor(private services: OtherServices) {}

  /**
   * Gets a flat list of clients for the specified potential
   * affiliate.
   */
  async list({ ssoID }: BySSOIDLookup): Promise<ClientDTO[]> {
    const clients = await Client.findAll({
      where: {
        statusID: {
          [Op.eq]: ClientStatusID.ACTIVE,
        },
        referralType: {
          [Op.eq]: ClientReferralType.AFFILIATE,
        },
      },
      include: [
        {
          model: AffiliateContact,
          required: true,
          attributes: [],
          where: {
            ssoID: {
              [Op.eq]: ssoID,
            },
          },
        },
        {
          model: ClientPhoneNumber,
          separate: true,
          order: [["cc_id", "ASC"]],
        },
      ],
      order: [
        ["time_added", "DESC"],
        ["clientid", "DESC"],
      ],
    });

    return clients.map((client: Client) => this.serialize(client));
  }

  /**
   * Gets an existing client by name, phone number or email.
   *
   * This is used to stop users from creating a client that another user
   * already has. This is _not_ to be used to avoid inserting duplicate
   * records. This is NOT CONCURRENCY SAFE.
   */
  async find({ email, phoneNumber }: ByNamePhoneNumberOrEmailLookup): Promise<ClientDTO | null> {
    const filters: WhereOptions[] = [];
    const phoneNumberFilters: WhereOptions[] = [];

    // e-mail addresses are not case sensitive so we lower-case on both
    // ends to simulate a case-insensitive search.. we can't use ILIKE
    // because its not supported by older MySQL versions or SQLite
    if (email) {
      const sanitizedEmail = email.toLowerCase().trim();
      filters.push(Sequelize.where(Sequelize.fn("lower", Sequelize.col("email")), sanitizedEmail));
    }

    if (phoneNumber) {
      const sanitizedPhoneNumber = this.sanitizePhoneNumber(phoneNumber);

      phoneNumberFilters.push({
        value: {
          [Op.eq]: sanitizedPhoneNumber,
        },
      });
    }

    const phoneNumbers = await ClientPhoneNumber.findAll({
      where: {
        [Op.or]: phoneNumberFilters,
      },
    });

    if (phoneNumbers) {
      filters.push({
        id: {
          [Op.in]: phoneNumbers.map((entry: ClientPhoneNumber) => entry.clientID),
        },
      });
    }

    const client = await Client.findOne({
      where: {
        [Op.or]: filters,
      },
      include: [
        {
          model: ClientPhoneNumber,
          order: [["cc_id", "ASC"]],
          separate: true,
        },
      ],
    });

    if (!client) {
      return null;
    }

    return this.serialize(client);
  }

  /**
   * Creates a new client that is owned by the specified {@paramref owner}.
   *
   * This will only work if the owner is a verified affiliate and will
   * fail with {@see AppError} if its not.
   */
  async create({ ssoID }: BySSOIDLookup, dto: CreateClientDTO): Promise<ClientDTO> {
    const { name, email, address, locationID, comment, phoneNumbers, source } = dto;

    const [owner, countryID] = await Promise.all([
      this.services.me.owner({ ssoID }),
      // lookup countryID because countries are included
      // in the location tree, we don't expose countryID to
      // the outside world, but we need to set it
      // for backwards compatibility reasons
      locationID ? this.services.locations.countryID({ locationID }) : Promise.resolve(null),
    ]);

    if (!owner) {
      throw new AppError("Current user has no owner", {
        statusCode: 400,
        extra: {
          ssoID,
        },
      });
    }

    if (locationID && !countryID) {
      throw new AppError("Cannot find location or location's country", {
        statusCode: 400,
        extra: {
          locationID,
        },
      });
    }

    // setting this for backwards compatibility, phone numbers
    // are stored separately as a list in zn_pf_client_cells
    const phoneNumber = dto.phoneNumbers[0]?.value || null;

    // make sure the user does not take ownership of a client
    // that somebody else already entered into the database
    // this is a business requirement, not a technical solution
    // to preventing duplicates (we have unique constraints for that)
    const existingClient = await this.find({ email, phoneNumber });
    if (existingClient) {
      throw new AppError("Client already exists", {
        statusCode: 409,
        extra: {
          name,
          email,
          phoneNumber,
        },
      });
    }

    const client = await Client.create(
      {
        statusID: ClientStatusID.ACTIVE,
        source: source || SourcePlatform.AFFILIATES_MOBILE_APP,
        userID: owner.userID,
        agencyID: owner.agencyID,
        name,
        email,
        address,
        countryID,
        locationID,
        comment,
        referralType: ClientReferralType.AFFILIATE,
        referralID: owner.affiliateID,
        ClientPhoneNumbers: phoneNumbers.map((phoneNumber: CreateClientPhoneNumberDTO) => ({
          value: phoneNumber.value,
          // add the short version too because the staff app duplication check is based on the short version
          shortValue: phoneNumber.value.slice(-9),
        })),
      },
      {
        include: [ClientPhoneNumber],
      }
    );

    return this.serialize(client);
  }

  /**
   * Serializes a queried {@see Client} model instance into a DTO.
   */
  private serialize(client: Client): ClientDTO {
    return {
      id: client.id,
      name: client.name,
      email: client.email || null,
      phoneNumbers: client.ClientPhoneNumbers.map((phoneNumber: ClientPhoneNumber) => ({
        id: phoneNumber.id,
        value: phoneNumber.value,
      })),
      address: client.address || null,
      locationID: client.locationID || null,
      comment: client.comment || null,
      createdAt: client.createdAt.toISOString(),
      updatedAt: client.updatedAt.toISOString(),
    };
  }

  /**
   * Sanitizes the specified phone number into a valid E.164 phone number.
   */
  private sanitizePhoneNumber(value: string): string {
    const errorMessage = `${value} is not a valid phone number`;
    const errorData = { statusCode: 400 };

    let parsedPhoneNumber: PhoneNumber | undefined | null = null;

    try {
      parsedPhoneNumber = parsePhoneNumberFromString(value);
    } catch (error) {
      throw new AppError(errorMessage, { ...errorData, error });
    }

    if (!parsedPhoneNumber) {
      throw new AppError(errorMessage, errorData);
    }

    return parsedPhoneNumber.number as string;
  }
}

export default ClientsService;
