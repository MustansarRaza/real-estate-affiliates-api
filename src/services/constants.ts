export enum ErrorMessages {
  USER_AFFILIATE_ERROR_MESSAGE = "Current user is not an affiliate",
}

export const FORMAT_DATE = {
  FULL_DATE_24_HOUR_FORMAT_DATE_FNS: "YYYY-MM-DD HH:mm:ss",
};
