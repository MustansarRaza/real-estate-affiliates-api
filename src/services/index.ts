import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
