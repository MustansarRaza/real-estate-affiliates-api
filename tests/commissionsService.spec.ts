import { AffiliateCommissionType } from "@pf/taxonomies";
import { CommissionsService } from "affiliates-api/services";
import each from "jest-each";
import MockDate from "mockdate";
import { loadSequelizeFixtures } from "./support";
import { createAffiliateCommission, ModelFixtureNames } from "./support/data";

describe("CommissionsService", () => {
  const service = new CommissionsService();

  beforeEach(async () => {
    await loadSequelizeFixtures(
      ModelFixtureNames.COUNTRIES,
      ModelFixtureNames.LOCATIONS,
      ModelFixtureNames.ADDRESSES,
      ModelFixtureNames.CATEGORIES,
      ModelFixtureNames.PROJECTS,
      ModelFixtureNames.AFFILIATE_COMMISSION_PROPERTY_TYPE
    );
    jest.resetAllMocks();
  });

  afterEach(() => {
    MockDate.reset();
  });

  describe("getDetailsFor", () => {
    it("should only get the details for the specified projectIDs and affiliateTypeId", async () => {
      const projectId1 = 1;
      const projectId2 = 2;

      const affiliateCommission1 = await createAffiliateCommission({
        projectId: projectId1,
      });
      await createAffiliateCommission({
        projectId: projectId2,
      });

      const affiliateCommissions = await service.getDetailsFor({
        projectIDs: [projectId1],
        affiliateTypeId: 1,
      });

      expect(Object.keys(affiliateCommissions)).toHaveLength(1);
      expect(affiliateCommissions[projectId1]).toHaveLength(1);
      expect(affiliateCommissions[projectId1][0].projectId).toBe(affiliateCommission1.projectId);
    });

    it("should only get the details for the specified filter options", async () => {
      const projectId1 = 1;
      const projectId2 = 2;

      const affiliateCommission1 = await createAffiliateCommission({
        projectId: projectId1,
      });
      await createAffiliateCommission({
        projectId: projectId2,
        propertyTypeId: 2,
      });

      const affiliateCommissions = await service.getDetailsFor({
        projectIDs: [1, 2],
        affiliateTypeId: 1,
        propertyTypeId: 1,
        createdBy: 483,
        commissionType: AffiliateCommissionType.DISCOUNT_MODEL,
      });

      expect(Object.keys(affiliateCommissions)).toHaveLength(1);
      expect(affiliateCommissions[projectId1]).toHaveLength(1);
      expect(affiliateCommissions[projectId1][0].projectId).toBe(affiliateCommission1.projectId);
      expect(affiliateCommissions[projectId1][0].propertyTypeId).toBe(1);
      expect(affiliateCommissions[projectId1][0].affiliateTypeId).toBe(1);
      expect(affiliateCommissions[projectId1][0].createdBy).toBe(483);
      expect(affiliateCommissions[projectId1][0].commissionType).toBe(
        AffiliateCommissionType.DISCOUNT_MODEL
      );
    });

    each([
      ["2021-04-30T00:00:00.000Z", null, null],
      ["2021-04-30T00:00:00.000Z", "2021-01-27T19:00:00.000Z", null],
      ["2021-04-30T00:00:00.000Z", null, "2022-01-26T19:00:00.000Z"],
      ["2021-04-30T00:00:00.000Z", "2021-01-27T19:00:00.000Z", "2022-01-26T19:00:00.000Z"],
    ]).it(
      "should only get the details for the commmissions against the project that satisfies the condition that current date %s lies within the startDate %s and endDate %s",
      async (currentDate, startDate, endDate) => {
        MockDate.set(currentDate);
        const projectId1 = 1;

        const affiliateCommission1 = await createAffiliateCommission({
          projectId: projectId1,
          startDate: startDate,
          endDate: endDate,
        });

        const affiliateCommissions = await service.getDetailsFor({
          projectIDs: [projectId1],
          affiliateTypeId: 1,
        });

        expect(Object.keys(affiliateCommissions)).toHaveLength(1);
        expect(affiliateCommissions[projectId1]).toHaveLength(1);
        expect(affiliateCommissions[projectId1][0].projectId).toBe(affiliateCommission1.projectId);
      }
    );

    each([
      ["2021-01-01T00:00:00.000Z", "2021-01-27T19:00:00.000Z", null],
      ["2022-02-27T00:00:00.000Z", null, "2022-01-26T19:00:00.000Z"],
      ["2021-01-01T00:00:00.000Z", "2021-01-27T19:00:00.000Z", "2022-01-26T19:00:00.000Z"],
      ["2022-02-27T00:00:00.000Z", "2021-01-27T19:00:00.000Z", "2022-01-26T19:00:00.000Z"],
    ]).it(
      "should not get the details for the commmissions against the project that doesn't satisfies the condition that current date %s lies within the startDate %s and endDate %s",
      async (currentDate, startDate, endDate) => {
        MockDate.set(currentDate);
        const projectId1 = 1;

        await createAffiliateCommission({
          projectId: projectId1,
          startDate: startDate,
          endDate: endDate,
        });

        const affiliateCommissions = await service.getDetailsFor({
          projectIDs: [projectId1],
          affiliateTypeId: 1,
        });

        expect(Object.keys(affiliateCommissions)).toHaveLength(0);
      }
    );
  });

  describe("getCommissionsBreakdownForAffiliate", () => {
    it("return the affiliate commission breakdown against the projectIds and affiliate type", async () => {
      MockDate.set("2021-04-30T00:00:00.000Z");

      const projectId1 = 1;
      const projectId2 = 2;
      const projectId3 = 3;

      await createAffiliateCommission({
        projectId: projectId1,
        affiliateTypeId: 1,
        commissionType: AffiliateCommissionType.FLAT_COMMISSION,
        propertyTypeId: 1,
        commission: 2.5,
        createdAt: "2021-04-30T00:00:00.000Z",
        updtaedAt: "2021-04-30T00:00:00.000Z",
      });
      await createAffiliateCommission({
        projectId: projectId2,
        affiliateTypeId: 1,
        commissionType: AffiliateCommissionType.DISCOUNT_MODEL,
        propertyTypeId: 2,
        downPayment: 30,
        sellingPrice: 91,
        commission: 2,
        createdAt: "2021-04-30T00:00:00.000Z",
        updtaedAt: "2021-04-30T00:00:00.000Z",
      });
      await createAffiliateCommission({
        projectId: projectId2,
        affiliateTypeId: 1,
        commissionType: AffiliateCommissionType.DISCOUNT_MODEL,
        propertyTypeId: 2,
        downPayment: 30,
        sellingPrice: 93,
        commission: 3,
        createdAt: "2021-04-30T00:00:00.000Z",
        updtaedAt: "2021-04-30T00:00:00.000Z",
      });
      await createAffiliateCommission({
        projectId: projectId3,
        affiliateTypeId: 1,
        commissionType: AffiliateCommissionType.UNIT_FROM,
        propertyTypeId: 3,
        unitFrom: 1,
        unitTo: 4,
        commission: 3,
        createdAt: "2021-04-30T00:00:00.000Z",
        updtaedAt: "2021-04-30T00:00:00.000Z",
      });
      await createAffiliateCommission({
        projectId: projectId3,
        affiliateTypeId: 1,
        commissionType: AffiliateCommissionType.UNIT_FROM,
        propertyTypeId: 4,
        unitFrom: 5,
        unitTo: 1000,
        commission: 4,
        createdAt: "2021-04-30T00:00:00.000Z",
        updtaedAt: "2021-04-30T00:00:00.000Z",
      });

      const affiliateCommissionBreakdown = await service.getCommissionsBreakdownForAffiliate({
        projectIDs: [projectId1, projectId2, projectId3],
        affiliateTypeId: 1,
      });

      expect(Object.keys(affiliateCommissionBreakdown)).toHaveLength(3);
      expect(affiliateCommissionBreakdown).toMatchSnapshot();
    });

    it("returns empty commission breakdown mapping if there is no commission records against any of the projectId and affiliate type", async () => {
      const projectId1 = 1;
      const projectId2 = 2;

      const affiliateCommissionBreakdown = await service.getCommissionsBreakdownForAffiliate({
        projectIDs: [projectId1, projectId2],
        affiliateTypeId: 1,
      });

      expect(Object.keys(affiliateCommissionBreakdown)).toHaveLength(0);
    });
  });

  describe("getProjectCommissionStats", () => {
    it("return the commission stats against the projectIds", async () => {
      const projectId1 = 1;
      const projectId2 = 2;

      await createAffiliateCommission({ projectId: projectId1, commission: 2.5 });
      await createAffiliateCommission({ projectId: projectId1, commission: 4 });
      await createAffiliateCommission({ projectId: projectId2, commission: 3.5 });

      const projectCommissionStats = await service.getProjectCommissionStats({
        projectIDs: [projectId1, projectId2],
        affiliateTypeId: 1,
      });

      expect(projectCommissionStats.size).toBe(2);
      expect(projectCommissionStats.get(projectId1)).toStrictEqual({
        max: 4,
        min: 2.5,
      });
      expect(projectCommissionStats.get(projectId2)).toStrictEqual({
        max: 3.5,
        min: 3.5,
      });
    });

    it("includes only those commissions for calculating stats matching with the specified filter options", async () => {
      const projectId1 = 1;

      await createAffiliateCommission({ projectId: projectId1, commission: 2.5 });
      await createAffiliateCommission({ projectId: projectId1, commission: 4, propertyTypeId: 3 });
      await createAffiliateCommission({ projectId: projectId1, commission: 3.5 });

      const projectCommissionStats = await service.getProjectCommissionStats({
        projectIDs: [projectId1],
        affiliateTypeId: 1,
        propertyTypeId: 1,
      });

      expect(projectCommissionStats.size).toBe(1);
      expect(projectCommissionStats.get(projectId1)).toStrictEqual({
        max: 3.5,
        min: 2.5,
      });
    });
  });
});
