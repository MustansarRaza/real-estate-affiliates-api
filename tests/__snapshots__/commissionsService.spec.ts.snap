// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommissionsService getCommissionsBreakdownForAffiliate return the affiliate commission breakdown against the projectIds and affiliate type 1`] = `
Object {
  "1": Object {
    "flat_commission": Object {
      "1": Array [
        Object {
          "additionalCommission": null,
          "affiliateTypeId": 1,
          "classifiedCredits": null,
          "commission": 2.5,
          "commissionType": "flat_commission",
          "createdAt": "2021-04-30 00:00:00",
          "createdBy": 483,
          "downPayment": 30,
          "endDate": null,
          "id": 1,
          "inventorySizeFrom": null,
          "inventorySizeTo": null,
          "inventoryType": null,
          "projectId": 1,
          "propertyTypeId": 1,
          "sellingPrice": null,
          "soldUnitsMax": null,
          "soldUnitsMin": null,
          "startDate": null,
          "totalCommission": null,
          "updatedAt": "2021-04-30 00:00:00",
        },
      ],
    },
  },
  "2": Object {
    "discount_model": Object {
      "2": Array [
        Object {
          "additionalCommission": null,
          "affiliateTypeId": 1,
          "classifiedCredits": null,
          "commission": 2,
          "commissionType": "discount_model",
          "createdAt": "2021-04-30 00:00:00",
          "createdBy": 483,
          "downPayment": 30,
          "endDate": null,
          "id": 2,
          "inventorySizeFrom": null,
          "inventorySizeTo": null,
          "inventoryType": null,
          "projectId": 2,
          "propertyTypeId": 2,
          "sellingPrice": 91,
          "soldUnitsMax": null,
          "soldUnitsMin": null,
          "startDate": null,
          "totalCommission": null,
          "updatedAt": "2021-04-30 00:00:00",
        },
        Object {
          "additionalCommission": null,
          "affiliateTypeId": 1,
          "classifiedCredits": null,
          "commission": 3,
          "commissionType": "discount_model",
          "createdAt": "2021-04-30 00:00:00",
          "createdBy": 483,
          "downPayment": 30,
          "endDate": null,
          "id": 3,
          "inventorySizeFrom": null,
          "inventorySizeTo": null,
          "inventoryType": null,
          "projectId": 2,
          "propertyTypeId": 2,
          "sellingPrice": 93,
          "soldUnitsMax": null,
          "soldUnitsMin": null,
          "startDate": null,
          "totalCommission": null,
          "updatedAt": "2021-04-30 00:00:00",
        },
      ],
    },
  },
  "3": Object {
    "unit_sold": Object {
      "3": Array [
        Object {
          "additionalCommission": null,
          "affiliateTypeId": 1,
          "classifiedCredits": null,
          "commission": 3,
          "commissionType": "unit_sold",
          "createdAt": "2021-04-30 00:00:00",
          "createdBy": 483,
          "downPayment": 30,
          "endDate": null,
          "id": 4,
          "inventorySizeFrom": null,
          "inventorySizeTo": null,
          "inventoryType": null,
          "projectId": 3,
          "propertyTypeId": 3,
          "sellingPrice": null,
          "soldUnitsMax": null,
          "soldUnitsMin": null,
          "startDate": null,
          "totalCommission": null,
          "updatedAt": "2021-04-30 00:00:00",
        },
      ],
      "4": Array [
        Object {
          "additionalCommission": null,
          "affiliateTypeId": 1,
          "classifiedCredits": null,
          "commission": 4,
          "commissionType": "unit_sold",
          "createdAt": "2021-04-30 00:00:00",
          "createdBy": 483,
          "downPayment": 30,
          "endDate": null,
          "id": 5,
          "inventorySizeFrom": null,
          "inventorySizeTo": null,
          "inventoryType": null,
          "projectId": 3,
          "propertyTypeId": 4,
          "sellingPrice": null,
          "soldUnitsMax": null,
          "soldUnitsMin": null,
          "startDate": null,
          "totalCommission": null,
          "updatedAt": "2021-04-30 00:00:00",
        },
      ],
    },
  },
}
`;
